tritium-metrics/src/main/java/com/palantir/tritium/metrics/Reservoirs.java

```diff
@@ -47,9 +49,17 @@ private Reservoirs() {
         return hdrHistogramReservoirSupplier(twoSignificantDigits());
     }
 
+    static Supplier<Reservoir> slidingTimeWindowArrayReservoir(long window, TimeUnit timeUnit) {
+        return slidingTimeWindowArrayReservoirSupplier(window, timeUnit);
+    }
+
     static Supplier<Reservoir> hdrHistogramReservoirSupplier(Supplier<Recorder> recorder) {
         checkNotNull(recorder, "recorder");
         return () -> new HdrHistogramReservoir(recorder.get());
     }
 
+    static Supplier<Reservoir> slidingTimeWindowArrayReservoirSupplier(long window, TimeUnit timeUnit) {
+        return () -> new SlidingTimeWindowArrayReservoir(window, timeUnit);
```

Comment:
I'm leaning toward this `Reservoir` class not actually returning `Supplier<Reservoir>` but just `Reservoir` and having caller handle either using the method reference `createWithReservoirType(Reservoirs::hdrHistogramReservoir);` or lambda `createWithReservoirType(() -> Reservoirs.slidingTimeWindow(window, windowUnit));`.

```
    /**
     * Supplies reservoirs backed by high dynamic range histograms that store measurements for the duration of the
     * reservoir's lifetime.
     *
     * <p>
     * See <a href="http://taint.org/2014/01/16/145944a.html">
     * Don’t use Timers with exponentially-decaying reservoirs in Graphite
     * </a>
     */
    @Nonnull
    static Reservoir hdrHistogramReservoir() {
        Recorder recorder = twoSignificantDigits().get();
        return hdrHistogramReservoir(recorder);
    }

    @Nonnull
    private static Reservoir hdrHistogramReservoir(Recorder recorder) {
        checkNotNull(recorder, "recorder");
        return new HdrHistogramReservoir(recorder);
    }

    /**
     * Supplies reservoirs backed by sliding time window array that store measurements for the specified sliding
     * time window.
     *
     * <p>
     * See also:
     * <ul>
     * <li>
     * <a href="https://github.com/dropwizard/metrics/issues/1138">
     * Drop-in replacement of sliding time window reservoir
     * </a>
     * </li>
     * <a href="https://github.com/dropwizard/metrics/pull/1139">
     * Issue #1138 Replacement of sliding time window
     * </a>
     * <li>
     * <a href="https://medium.com/hotels-com-technology/your-latency-metrics-could-be-misleading-you-how-hdrhistogram-can-help-9d545b598374">
     * Your Dropwizard Latency Metrics Could Be Misleading You — How Rolling-Metrics and HdrHistogram Can Help
     * </a>
     * </li>
     * </ul>
     * </p>
     *
     * @param window window of time
     * @param windowUnit unit for window
     * @return reservoir
     */
    @Nonnull
    static Reservoir slidingTimeWindow(long window, TimeUnit windowUnit) {
        return slidingTimeWindow(window, windowUnit, Clock.defaultClock());
    }

    @Nonnull
    @VisibleForTesting
    static Reservoir slidingTimeWindow(long window, TimeUnit windowUnit, Clock clock) {
        checkNotNull(windowUnit, "windowUnit");
        checkNotNull(clock, "clock");
        return new SlidingTimeWindowArrayReservoir(window, windowUnit, clock);
    }
```

Reply:
ack